src/com/seafile/seadroid2/transfer/DownloadTaskManager.java

```diff
@@ -0,0 +1,250 @@
+package com.seafile.seadroid2.transfer;
+
+import android.util.Log;
+import com.google.common.collect.Lists;
+import com.seafile.seadroid2.ConcurrentAsyncTask;
+import com.seafile.seadroid2.account.Account;
+import com.seafile.seadroid2.util.Utils;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Download task manager
+ *
+ * Created by Logan on 15/2/4.
+ */
+public class DownloadTaskManager extends TransferManager {
+    private static final String DEBUG_TAG = "DownloadTaskManager";
+
+    /** contains all download tasks, including failed, cancelled, finished, downloading, waiting tasks. */
+    private ArrayList<DownloadTask> allTaskList = Lists.newArrayList();
+    /** contains currently downloading tasks */
+    private List<DownloadTask> downloadingList = Lists.newArrayList();
+    /** contains waiting for download tasks */
+    private List<DownloadTask> waitingList = Lists.newArrayList();
+
+    private static final int DOWNLOAD_MAX_COUNT = 2;
+
+    /**
+     * Add a new download task.
+     * call this method to execute a task instantly.
+     */
+    public int addTask(Account account,
+                       String repoName,
+                       String repoID,
+                       String path) {
+        Iterator<DownloadTask> iter = allTaskList.iterator();
+        while (iter.hasNext()) {
+            DownloadTask task = iter.next();
+            if (task.getRepoID().equals(repoID) && task.getPath().equals(path)) {
+                if (task.getState() == TaskState.CANCELLED || task.getState() == TaskState.FAILED
+                        || task.getState() == TaskState.FINISHED) {
+                    // If there is a duplicate, but it has failed or been
+                    // cancelled, remove it first
+                    iter.remove();
+                    break;
+                } else {
+                    // A duplicate task is downloading
+                    return task.getTaskID();
+                }
+            }
+        }
+
+        DownloadTask task = new DownloadTask(++notificationID, account, repoName, repoID, path, listener);
+        allTaskList.add(task);
+        ConcurrentAsyncTask.execute(task);
+        return task.getTaskID();
+    }
+
+    public int addTaskToQue(Account account,
+                            String repoName,
+                            String repoID,
+                            String path) {
+        int taskID = -1;
+        synchronized (waitingList) {
```

Comment:
We access both `allTasksList` and `waitlingList` in this method，so it's not enough only use `synchronized` on `waitingList`, we can simply add `synchronized` to the whole method:

```java
public synchronized int addTaskToQue ...
```

This also applies to the `remove` method below (and `UploadTaskManagner`)